### PR TITLE
ENH: Add CPU feature detection for POWER10 (VSX4)

### PR DESCRIPTION
**Pull request by rafaelcfsousa**
This PR adds CPU feature detection for POWER10 (VSX4). It will allow us to accelerate some NumPy operations, e.g., floor_divide, using the new SIMD builtins available on POWER10.

This PR follows the instructions presented here: #19537

### CODE REVIEW

File: numpy/distutils/ccompiler_opt.py [hunk 1]
```diff
@@ -294,6 +294,8 @@ class _Config:
         VSX2 = dict(interest=2, implies="VSX", implies_detect=False),
         ## Power9/ISA 3.00
         VSX3 = dict(interest=3, implies="VSX2", implies_detect=False),
+        ## Power10/ISA 3.1
+        VSX4 = dict(interest=4, implies="VSX3", extra_checks="VSX4_MMA", implies_detect=False),
```
Comment (mattip) [inline numpy/distutils/ccompiler_opt.py:298]: Make linter happy
```suggestion
        VSX4 = dict(interest=4, implies="VSX3", extra_checks="VSX4_MMA",
                    implies_detect=False),
```

**Comment by mattip** [comment]
Does this enable a speed bump in benchmarks for that architecture?

**Comment by rafaelcfsousa** [comment]
> Does this enable a speed bump in benchmarks for that architecture?

Hi @mattip. Yes, it does. As an example, we have speedup between 15-25% for floor_divide (unsigned types). We have something similar for the operation 'remainder'. But notice that only with the modifications of this PR the optimizations described above are not enabled. My plan is to submit other PRs separately with the optimizations I am working on NumPy for the POWER architectures (POWER8, POWER9, and POWER10). Thanks.

File: numpy/distutils/ccompiler_opt.py (hunk 1 above)
Comment (rafaelcfsousa) [reply numpy/distutils/ccompiler_opt.py]: Done!, Thanks!

**Comment by mattip** [comment]
Could you add this to the documentation [here](https://github.com/numpy/numpy/blob/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/source/reference/simd/build-options.rst?plain=1#L164) (VSX3 is also missing). Also documentation is missing in the [rendered table](https://24459-908607-gh.circle-artifacts.com/0/doc/build/html/reference/simd/build-options.html?highlight=vsx3#on-ibm-power-big-endian) which is generate by [this code](https://github.com/numpy/numpy/blob/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/source/reference/simd/gen_features.py#L180), I wonder why VSX4 does not show up there. Is something missing for Power10?

**Comment by mattip** [comment]
This should probably also get an [enhancement release note](https://github.com/numpy/numpy/tree/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/release/upcoming_changes#changelog)

**Comment by rafaelcfsousa** [comment]
> Could you add this to the documentation [here](https://github.com/numpy/numpy/blob/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/source/reference/simd/build-options.rst?plain=1#L164) (VSX3 is also missing). 

The table you pointed out shows the minimum CPU features (MIN) that can safely run on a wide range of platforms.  For POWER architecture, VSX2/Power8 is defined as the minimum CPU features (MIN), which means that the documentation is correct, thus no modification is required.  See below the MIN that is defined for POWER and also for the other architectures (defined here: [link](https://github.com/numpy/numpy/blob/78c6713b4ee06bec1fa579f94b466cbe5fe4e8f3/numpy/distutils/ccompiler_opt.py#L230)).

``` 
conf_min_features = dict( 
        x86 = "SSE SSE2",
        x64 = "SSE SSE2 SSE3",
        ppc64 = '', # play it safe
        ppc64le = "VSX VSX2",
        s390x = '',
        armhf = '', # play it safe
        aarch64 = "NEON NEON_FP16 NEON_VFPV4 ASIMD"
    )
```



> Also documentation is missing in the [rendered table](https://24459-908607-gh.circle-artifacts.com/0/doc/build/html/reference/simd/build-options.html?highlight=vsx3#on-ibm-power-big-endian) which is generate by [this code](https://github.com/numpy/numpy/blob/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/source/reference/simd/gen_features.py#L180), I wonder why VSX4 does not show up there. Is something missing for Power10?

VSX4 does not show up there because I did not execute the script you referred :-). After executing the script, I see VSX4 (Power10) added to the table.  I just pushed a commit with the changes of the script to update the table. Notice that the same was done in this PR (#20552). 

Still related to PR #20552 I found some additional tests there that I also included in this PR (see here: [link](https://github.com/rafaelcfsousa/numpy/commit/3784cac6f79b1a2f5828957173cac81bbc483ba3)).

> This should probably also get an [enhancement release note](https://github.com/numpy/numpy/tree/4b985e2b85c7efc1d976835f74c3e45315c5ae28/doc/release/upcoming_changes#changelog)

This is a great idea! But I think that it would be more valuable to add this within 3~4 weeks as I will be adding optimizations to accelerate some operations for VSX4/Power10. Is it ok if I proceed this way?

Thanks again!

**Comment by mattip** [comment]
> Is it ok if I proceed this way?

Just to be sure I opened #20849 with a 1.23 milestone. Now we have to remember to close it when you submit the next PR :)

**Comment by mattip** [comment]
Thanks @rafaelcfsousa

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @rafaelcfsousa! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute/. 
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

**Pull request merged**